avro-fastserde/src/main/java11/com/linkedin/avro/fastserde/BufferBackedPrimitiveFloatList.java

```diff
@@ -0,0 +1,384 @@
+ * - It does not, by default, box primitive floats into Object Floats, though it will still do so if the
+ *   regular functions are called (e.g.: {@link #get(int)}, for compatibility purposes. In order to avoid
+ *   boxing, the {@link #getPrimitive(int)} function can be used instead.
+
+ * - It does not maintain a reference to a {@link Schema} instance, since that schema would always be the
+ *   same. Instead, it defines a static {@link #SCHEMA} which is used by all instances.
+
+ * - It re-implements {@link #compareTo(GenericArray)}, {@link #equals(Object)} and {@link #hashCode()}
+ *   in order to leverage the primitive types, rather than causing unintended boxing.
+
+ *   Using VarHandle(JDK9+ API) to speed up float-array deserialization: We allocate byte array to store the raw bytes
+ *   from BinaryDecoder and deserialize them only during array element access. We cache the results into the elements
+ *   array after the first get access of the array so that sub-sequent array access are fast.
+ *   TODO: Provide arrays for other primitive types.
+ */
+public class BufferBackedPrimitiveFloatList extends AbstractList<Float>
+    implements GenericArray<Float>, Comparable<GenericArray<Float>>, PrimitiveFloatList {
+  private static final float[] EMPTY = new float[0];
+  private static final int FLOAT_SIZE = Float.BYTES;
+  private static final Schema FLOAT_SCHEMA = Schema.create(Schema.Type.FLOAT);
+  private static final Schema SCHEMA = Schema.createArray(FLOAT_SCHEMA);
+  private int size;
+  private float[] elements = EMPTY;
+  private boolean isCached = false;
+  private byte[] byteBuffer;
+
+  private static final VarHandle VH = MethodHandles.byteArrayViewVarHandle(float[].class, ByteOrder.LITTLE_ENDIAN);
+
+  public BufferBackedPrimitiveFloatList(int capacity) {
+    if (capacity != 0) {
+      elements = new float[capacity];
+    }
+  }
+
+  public BufferBackedPrimitiveFloatList(Collection<Float> c) {
+    if (c != null) {
+      elements = new float[c.size()];
+      addAll(c);
+    }
+  }
+
+  /**
+   * Instantiate (or re-use) and populate a {@link BufferBackedPrimitiveFloatList} from a {@link Decoder}.
+
+   * N.B.: the caller must ensure the data is of the appropriate type by calling {@link #isFloatArray(Schema)}.
+   *
+   * @param old old {@link BufferBackedPrimitiveFloatList} to reuse
+   * @param in {@link Decoder} to read new list from
+   * @return a {@link BufferBackedPrimitiveFloatList} with data, possibly the old argument reused
+   * @throws IOException on io errors
+   */
+  public static Object readPrimitiveFloatArray(Object old, Decoder in) throws IOException {
+    long length = in.readArrayStart();
+    long totalLength = 0;
+
+    if (length > 0) {
+      BufferBackedPrimitiveFloatList array = (BufferBackedPrimitiveFloatList) newPrimitiveFloatArray(old);
+
+      do {
+        long byteSize = length * FLOAT_SIZE;
+        byte[] buffer = new byte[(int)byteSize];
+        in.readFixed(buffer, 0, (int)byteSize);
```

Comment:
I think it would be ideal not to instantiate this new buffer here, when we could instead keep reuse the one that came in from the `old` instance. I took a stab at it here is what I got:

```
  public BufferBackedPrimitiveFloatList(int capacity) {
    this.byteBuffer = new byte[capacity * FLOAT_SIZE];
    if (capacity != 0) {
      elements = new float[capacity];
    }
  }

  public BufferBackedPrimitiveFloatList(Collection<Float> c) {
    this.byteBuffer = new byte[c.size() * FLOAT_SIZE];
    if (c != null) {
      elements = new float[c.size()];
      addAll(c);
    }
  }

  public static Object readPrimitiveFloatArray(Object old, Decoder in) throws IOException {
    long length = in.readArrayStart();
    long totalLength = 0;
    BufferBackedPrimitiveFloatList array = (BufferBackedPrimitiveFloatList) newPrimitiveFloatArray(old, (int) length);

    if (length > 0) {
      // First block
      long blockByteSize = length * FLOAT_SIZE;
      if (array.byteBuffer.length < blockByteSize) {
        array.byteBuffer = new byte[(int)blockByteSize];
      }
      in.readFixed(array.byteBuffer, 0, (int)blockByteSize);
      array.size = length;
      length = in.arrayNext();

      // Subsequent blocks, if any
      while (length > 0) {
        blockByteSize = length * FLOAT_SIZE;
        int currentUsefulPortionOfByteBuffer = array.size * FLOAT_SIZE;
        if (array.byteBuffer.length - currentUsefulPortionOfByteBuffer >= blockByteSize) {
          in.readFixed(array.byteBuffer, array.size, (int)blockByteSize);
        } else {
          /*
           * This implies memory copy when there is more than one float array
           * However, another option will be to keep the list of byte arrays
           * and track positions similar to {@link CompositeByteBuffer}
           */
          byte[] buffer = new byte[(int)blockByteSize + currentUsefulPortionOfByteBuffer];
          System.arraycopy(array.byteBuffer, 0, buffer, 0, currentUsefulPortionOfByteBuffer);
          in.readFixed(buffer, currentUsefulPortionOfByteBuffer, (int)blockByteSize);
          array.byteBuffer = buffer;
        }
        array.size += length;
        length = in.arrayNext();
      }
    }
    return array;
  }

  private static Object newPrimitiveFloatArray(Object old, int length) {
    if (old instanceof BufferBackedPrimitiveFloatList) {
      BufferBackedPrimitiveFloatList oldFloatList = (BufferBackedPrimitiveFloatList) old;
      oldFloatList.isCached = false;
      oldFloatList.size = 0;
      return oldFloatList;
    } else {
      // Just a place holder, will set up the elements later.
      return new BufferBackedPrimitiveFloatList(length);
    }
  }
```

Haven't tried it, so IDK if it works. Might be buggy. The significant changes are:

1. The constructor leverages the `capacity` param to pre-allocate a `byte[]` of the right size. Thanks to this, I made the assumption that I don't need to do any null check in `readPrimitiveFloatArray`.
2. `readPrimitiveFloatArray` never allocates a new `byte[]` if the data contains just one block, and if the old array is big enough (common case in our workload).
3. If there are subsequent blocks, it also keeps reusing the old array as long as it's big enough, but it will allocate one array per block as long as the old array isn't big enough.
4. We return the old array (properly re-initialized) in cases where length is zero, rather than throwing it away and creating a new empty one. This would mean we hang on to instances if we have a mixed workload where some of the arrays are empty and others are populated.

Reply:
This could be optimization in terms of performance but opens a way for forged attacks if serialized data is sensitive. I mean that we can try to read floats left from previous use of the same byte array. Not sure if you guys care about this case.

Reply:
could reuse only if same size ? (in which case it gets completely overwritten?)

Reply:
No IIUC it should be able to reuse as long as the old object is equal or bigger as what the new object needs.

I don't think this is a nice-to-have, since the older implementation will do full reuse, from my understanding. So this new code will allocate significantly more than the previous one?